designs/PackageVulnerability/FlagVulnerablePackages.md

```diff
@@ -0,0 +1,147 @@
+
+### Stage 1 - MVP
+
+For the [MVP](https://en.wikipedia.org/wiki/Minimum_viable_product) of this feature, we'll focus on:
+
+* Enabling ***consumers** to be informed about known vulnerabilities before installation* of a package.
+* Enabling ***authors** to be informed about known vulnerabilities in their packages*.
+* Leveraging **existing, curated, and publicly available known vulnerabilities** affecting NuGet packages.
+
+This should help prevent *new* installs of packages that are known to be vulnerable, and trigger authors into providing a fix through a new package version on nuget.org.
+
+NuGet will not support self-reporting of vulnerabilities at this time and instead rely on an existing data provider of curated vulnerability information. For the MVP of this feature, all vulnerability data will be sourced from external systems. The initial focus is on leveraging [GitHub's Security Workflow](https://github.com/features/security) and integration with [GitHub's GraphQL API](https://developer.github.com/v4/). This allows us to build an initial end-to-end experience without having to deal with the complexities of curating vulnerability reports.
+
+#### ✔️ In scope (Stage 1)
+
+* For CLI:
+  * A CLI command to **list vulnerabilities** in a project/solution project graph
+
+* For Visual Studio:
+  * **Display vulnerability information** in the package graph on `Installed` tab of Package Manager Dialog (PMUI)
+
+* For nuget.org:
+  * **Automatically ingest and refresh NuGet package vulnerability data from GitHub**
+  * **Surface package vulnerability information on-demand**:
+    * as part of package metadata in `v3` registration blobs (`JSON`)
+    * by showing vulnerability information to package owners (only!) on package details page and version history
+  * Basic (e-mail) notification to notify package publishers for any new vulnerability detected in one or more of their packages
+
+#### ❌ Out of scope (Stage 1)
+
+Some enhancements to this feature are out of scope for the initial implementation of this feature because there are several unknowns still to be clarified. See also the *Open Questions* section below.
+
+* Enable `restore` to Inform or Warn (see Stage 2)
+* Expose vulnerability metadata in nuget.org *Search* service (not required for MVP)
+* Display vulnerability information for *transitive packages* in PMUI (additional UX complexity)
+* Expose vulnerability information *publicly* on nuget.org's package details page and version history (disclosure policies TBD)
+
+### Stage 2
+
+Stage 2 of this feature will focus on *highlighting known vulnerabilities when the package is already installed*. This involves introducing new validations at package `restore` time, and new ways of dealing with them.
+
+#### ✔️ In scope (Stage 2)
+
+* For CLI:
+  * For `dotnet restore`, vulnerability reporting is opt-out. By default vulnerabilities are reported during a restore operation.
```

Comment:
I think this was brought up when the original design and I don't want to block/try to answer it here, but given the registration approach of reporting, reporting vulnerabilities at restore will incur a performance cost. 

We'll need a detailed design when we do get to phase 2. Personally I think we sohuld call it out the same way we're calling out the transitive dependency UX challenges.

Reply:
Agreed. I'll make a note here given it is our eventual goal at what cost though? :)